enhancements/ingress/host-network-configurable-ports.md

```diff
@@ -0,0 +1,298 @@
+The cluster administrator can set the ingress controller's `spec.endpointPublishingStrategy.hostNetworkParameters`.
+
+#### As an OpenShift developer, I want to make sure Ingress Controller documentation is updated with this new feature concerning the HostNetwork strategy
+
+#### As an OpenShift developer, automated e2e tests cover the case of deploying 2 ingress controllers in HostNetwork strategy to the same nodes with different sets of ports, so that I can validate the feature
+
+#### As a Cluster Administrator, I want the IngressController status to report problems provisioning the Ingress Controller pods in HostNetwork strategy when there is a port conflict
+
+### API Extensions
+
+This  proposal edits `IngressController.operator.openshift.io/v1`.
+
+### Operational Aspects of API Extensions
+
+N/A.
+
+### Implementation Details
+
+Implementing this needs changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+
+OpenShift Cluster Ingress Operator, creates a deployment for Router with environment variables for port bindings which OpenShift Router already respects:
+`ROUTER_SERVICE_HTTP_PORT`, `ROUTER_SERVICE_HTTPS_PORT`, `STATS_PORT`.
+
+Also the deployments should not have ip:port conflicts for `SNI_PORT` and `NO_SNI_PORT`. There are multiple ways to achieve this:
+
+- The cluster ingress operator configures the router to use a unique loopback address for the sni/nosni frontends. This loopback address is configured automatically, for example by hashing the ingresscontroller’s name.
+  As a result, the sni/nosni port numbers wouldn’t need to change to avoid the conflict as the routers would be using different IP addresses. This can be done with the introduction of two new environment variables for the router:`ROUTER_SERVICE_SNI_IP`, `ROUTER_SERVICE_SNI_IP`, and the configuration of the frontend to bind on this IPs.
+  The cluster ingress operator sets these environment variables for the router deployment from `127.0.0.1/8` CIDR based on the ingress controller name hash.
+- The router gets rid of the loopback hop using a Unix domain socket or other solution, which would have other advantages as well ([router's PR](https://github.com/openshift/router/pull/326)).
+
+### Risks and Mitigations
+
+#### Failure Modes
+
+N/A.
+
+#### Support Procedures
+n
+N/A.
+
+#### Port Conflict
+
+Users might define conflicting ports which would cause HAProxy process to fail at startup.
+
+A mitigation to this risk is to implement a validation feature in the reconciliation loop of the Cluster Ingress Operator
+to ensure all the ports defined in the `hostNetworkParameters` section are unique, and return an error with a meaningful message if there are conflicting ports.
+
+#### Usage of X-Forwarded headers
+
+The default forwarded header policy of the ingress controller is to append X-Forwarded headers. This results into the insertion of `X-Forwarded-Port` header on OpenShift router side (HAProxy acts as a reverse proxy).
```

Comment:
The default router binds on the standard 80, 443 ports which match the ports of the LB which "frontends" the router (`endpointPublishingStrategy == LoadBalancer`).
This works well for the applications like the image registry which needs to send the token retrieval URL back to the client. The token retrieval URL uses:
- OpenShift cluster's `*.apps` wildcard DNS which get resolved to the before mentioned LB
- original port (the one appended to `X-Forwarded-Port` by the router)

So, if the router is bound on a port `10443`  instead of `443`, the image registry will send the following URL:
```
https://registry.apps.<cluster_name>.<base_domain>.:10443/openshift/token?account=my-sa&client_id=docker&offline_token=true
```
which will fail as the "frontend" LB listens on the standard port `443` not `10443`.

Reply:
The default policy is `append`. So each host and port will be appended. The image registry should just use the first host and port in the headers. I don't think having non-standard behavior when one setting is changed is a good idea. It's confusing for a user who is trying to debug an issue.

Reply:
The image registry does take the first host and port, it's just that the first values are the router ones as the LB in front of the OpenShift cluster is L4 one, so it doesn't append any HTTP headers.

> I don't think having non-standard behavior when one setting is changed is a good idea. It's confusing for a user who is trying to debug an issue.

By "non-standard behavior" you mean the disabling of the append of the X-Forwarded headers? And by "one setting is changed" you mean the usage of the custom ports?
If so, the idea was to warn about this potential issue but keep the EP reasonably small moving the responsibility about using X-Forwarded headers to the applications. Do you have any idea of how to avoid the issue at reasonable cost?

Reply:
Discussed this offline. This mitigation is not to be implemented in the controller but can be applied by the user to workaround any issues which might arise from components which use the `X-Forwarded-Port` to reconstruct the request URL.